### PR TITLE
Multi export

### LINKED ISSUE #270: generateExportJson follow-ups
follow-ups to #254 :

* [x] `distortableCollection.avg_cm_per_pixel`
* [x] `generateExportJson` pulls from `distortableCollection.avg_cm_per_pixel`
* [x] startExport ajax call that gets the url for the `status.json` file
* [ ] refactor to get `_getStatusJson`  separated from outer ajax call `_runExport`
* [ ] `distortableCollection.startExport(callbackFn)` with basic `console.log` callback function to show progress every X seconds

### PR DESCRIPTION
**Pull request by sashadev-sky**
Fixes #270  (<=== Add issue number here)
/ PR at #310 

Also:
- #186 / #235 
- #29
- #62 
- Fix toolbar undefined #295 / issue at #273 
- Work towards #161 

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `grunt`
* [x] code is in uniquely-named feature branch and has no merge conflicts 📁
* [x] screenshots/GIFs are attached 📎 in case of UI updates
* [x] @mention the original creator of the issue in a comment below for help or for a review

> We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged** if your tests fail at first!

If tests do fail, click on the red `X` to learn why by reading the logs.

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/contributing-to-public-lab-software 

Thanks!

=================================

Changes:

- [ ] Removed:  `unused: "strict"` from JSHINT gruntfile
- [ ] `L.Map.BoxSelector` is fixed (Multi-select with shift-drag)
- [ ] `src/edit/tools` folder with `DistortableImage.ControlBar.js` and `DistortableImage.PopupBar.js`, each corresponding to the new class names for the 2 types of toolbars and `EditAction.js` the classname for the class we extend to create new actions.
- [ ] `leaflet-toolbar` upgraded to "0.4.0-alpha.2" and added as a dependency in package.json
- [ ] Two multi-image interface tools: `Exports`, `Deletes`
- [ ] Toolbar API: `removeTool`, `addTool`, `hasTool` for the `DistortableCollection`  and `DistortableImage.Edit` class - allows changing the toolbar actions in runtime
   - Did not use the implementation made by @rexagod in #155 but will circle back around later to review it and compare carefully 
- [ ] `DistortableCollection` and `DistortableImageOverlay` instantiation updates: new `action` option allows overwriting the default toolbar actions for respective toolbars initially if don't want to use the runtime API above
- [ ] `DistortableCollection` instantiation updates: images are now only added to the collection class, not the map directly. 
- [ ] `_updateCorners` and `_updateCorner()` are now `setCorners` and `setCorner()` but not ready to be documented yet - need `overlay.fire("update")` in them to work sep. but I don't want to change where we currently fire it in our code bc of the bugs with rotation, scaling ,etc.
- [ ] Deletion of unnecessary committed `leaflet.distortableimage.js.map` file two months ago
- [ ] Testing updates to match changes to `DistortableCollection` class

========================

Pending: 

- [ ] loader UI: considering for now just adding animated SVG loader icon to replace Export icon during load
- [ ] Multiple Exports still not working
- [ ] make a `DistortableCollection.Edit` class so the Tool API will work the same on it as on `DistortableImage.Edit` - then update docs

### CODE REVIEW

**Comment by sashadev-sky** [comment]
@jywarren Ok this is the new PR! Still have some documentation to move over from #310 but I still have not gotten the multiple exports to work! 

Big UI updates like a progress modal are difficult via ES5 because of implementation obstacles. I have never animated anything in svg before but I was thinking for now to try to animate a black loader circle that pops up in place of the Exports icon during loading. Then later circle back and add in all the 5 export steps and make it fancier, etc. 

Perhaps most important would be 1st fixing exports, merging this and connecting to MapKnitter. Currently have not yet implemented a way to pass callbacks for `startExport` but for now at least have the ability to remove tools, and then add new ones. Since actions are now a class we can recreate our own exports action in Mapknitter and add it.

**Comment by sashadev-sky** [comment]
The `removeTool`, `addTool`, `actions` API is working on the collection group and individual image instances. Here I limited the collection to just Export and img4 to just the 3 tools:
API

I documented it for the Collection class mostly and am going to hold off on completing docs because I think it would make sense if the Collection class had a connected `Edit` class like the overlay and then they can also share these functions

**Comment by jywarren** [comment]
Omg cool is your gh-pages demo running w exporting?? Rad Sasha!

**Comment by sashadev-sky** [comment]
@jywarren the exporting is still not working!! We can merge this on your approval because there is a lot going on and then try to fix the exporting in another PR. Or try to fix it here. If we get it working soon I can add a simple loader on to the page and try to connect it to MK by tomorrow

**Comment by sashadev-sky** [comment]
@rexagod @jywarren `BoxSelector` is fixed! I was right that Leaflet had fixed it in the #235 issue. @rexagod I think you may have also been correct about the `_mousedown` issue and that was a 2nd issue with it, fixed when I rewrote the way some of the event handlers in the  `DistortablCollection` class work. 

This gif shows how you can use it (precision looks perfect from bottom-right to top-left. Any other directions and it only works if you cover a lot more area - will open new issue to investigate that - not really a bug per say).

boxzoom

**Comment by jywarren** [comment]
Is this running on https://sasha.mapknitter.org ? So i can help debug? Thanks!!!

**Comment by jywarren** [comment]
On sasha.mk.org i see `/id/1562007112/status.json` in console when i click export multiple button with 2 selected. Then i see:

```
{"status_url":null,"status":"starting","tms":false,"geotiff":false,"zip":false,"jpg":false,"export_id":1562007112,"user_id":null,"size":null,"width":null,"height":null,"start_time":"2019-07-01 18:51:52 +0000","run_time":null,"gem_version":"1.0.7","cm_per_pixel":null}
```

Looking at this now.

**Comment by jywarren** [comment]
initial request to startExport contained: 

```
[{"id":143,"src":"http://sasha.mapknitter.org/examples/example.png","nodes":[{"lat":51.49089394353355,"lng":-0.1256154652219266},{"lat":51.49089394353355,"lng":-0.08561836439184846},{"lat":51.47090242798578,"lng":-0.13574348646216097},{"lat":51.47090242798578,"lng":-0.09574638563208283}],"cm_per_pixel":36.40625},{"id":176,"src":"http://sasha.mapknitter.org/examples/example.png","nodes":[{"lat":51.51,"lng":-0.07},{"lat":51.51,"lng":-0.03},{"lat":51.49,"lng":-0.08},{"lat":51.49,"lng":-0.04}],"cm_per_pixel":36.40625}]
```

**Comment by jywarren** [comment]
I'll try making a file `status.json` and manually uploading it to http://export.mapknitter.org

**Comment by jywarren** [comment]
OK, it uploaded fine initially, same, and then gave me http://export.mapknitter.org/id/1562007374/status.json

**Comment by jywarren** [comment]
Seems to be going OK...
```
{"status_url":"https://mapknitter-exports-warps.storage.googleapis.com/1562007374/status.json","status":"warping 1 of 2","tms":false,"geotiff":false,"zip":false,"jpg":false,"export_id":1562007374,"user_id":null,"size":null,"width":null,"height":null,"start_time":"2019-07-01 18:56:14 +0000","run_time":null,"gem_version":"1.0.7","cm_per_pixel":null}
```

**Comment by jywarren** [comment]
OK, well i'm not sure, but been digging a lot. I wonder if we're trying too ambitious a scale, with 3.x cm per pixel as default. Lots say: 

```
convert -contrast-stretch 0 public/warps/1562010305-working/w143- -crop 11911x11911+0+0! -flatten -distort Perspective '0,0 0,0 0,0 0,0 0,0 0,11911 0,0 0,11911' -flatten -crop 11911x11911+0+0! +repage public/warps/1562010305/w143.png
```

2 things there. First, `public/warps/1562010305-working/w143-` is not a real filename, not sure what's happening there. Second, `11911x11911+0+0` is a HUGE image. I'll try running this again with a lower scale.

**Comment by jywarren** [comment]
OK, for testing, i tried this: http://export.mapknitter.org/export?url=https://stable.publiclab.org/test.json&scale=30

with `cm_per_pixel` of 50, but i forgot to change the URL. Will try that too.

**Comment by jywarren** [comment]
http://export.mapknitter.org/id/1562007374/status.json without URL change...

**Comment by jywarren** [comment]
Well, that's looking better:

```
{"status_url":"https://mapknitter-exports-warps.storage.googleapis.com/1562007374/status.json","status":"compositing","tms":false,"geotiff":"public/warps/1562007374/1562007374.tif","zip":false,"jpg":false,"export_id":1562007374,"user_id":null,"size":null,"width":null,"height":null,"start_time":"2019-07-01 18:56:14 +0000","run_time":null,"gem_version":"1.0.7","cm_per_pixel":36.40625}
```

**Comment by jywarren** [comment]
Oops, i'm sorry, that last one was for an earlier export, and it's apparently compositing... it might just have been taking forever... 

Anyways this last one is /id/1562011515/status.json

**Comment by jywarren** [comment]
That's right and it's also running:

```
{"status_url":"https://mapknitter-exports-warps.storage.googleapis.com/1562011515/status.json","status":"compositing","tms":false,"geotiff":"public/warps/1562011515/1562011515.tif","zip":false,"jpg":false,"export_id":1562011515,"user_id":null,"size":null,"width":null,"height":null,"start_time":"2019-07-01 20:05:14 +0000","run_time":null,"gem_version":"1.0.7","cm_per_pixel":30.0}
```

**Comment by jywarren** [comment]
OK, the weird cutoff filename is from this line, but maybe it doesn't matter... not sure why it's not getting a real filename but if it's compositing, then maybe it's ok?

https://github.com/publiclab/mapknitter-exporter/blob/c767780392342c109b26a887971b3ab6bcd98c94/lib/mapknitterExporter.rb#L38

File: src/DistortableCollection.js [hunk 1]
```diff
@@ -243,9 +299,89 @@ L.DistortableCollection = L.FeatureGroup.extend({
...
+
+    // receives the URL of status.json, and starts running the updater to repeatedly fetch from status.json; 
+    // this may be overridden to integrate with any UI
+    function _defaultHandleStatusUrl(data) {
+      console.log(data);
+      statusUrl = "//export.mapknitter.org" + data;
+      opts.updater = opts.updater || _defaultUpdater;
+
+      // repeatedly fetch the status.json
+      updateInterval = setInterval(function intervalUpdater() {
```
Comment (jywarren) [inline src/DistortableCollection.js]: Here I realized we made a mistake and were running the interval /only/ on the display of status, but not actually fetching the status repeatedly. We were just showing the same value over and over without fetching a new copy from the server!

We need to rebuild files over this change (i made the change directly to your PR, hope that's OK!)

**Comment by jywarren** [comment]
OK, built too!

**Comment by jywarren** [comment]
Fixing #315  in this too!

**Comment by jywarren** [comment]
Testing... on my own gh-pages

image

**Comment by jywarren** [comment]
Aha - ok, I solved an issue where we hadn't been providing an `image_file_name` and fixed it both here and in the exporter project -- but now I see that our lat/lon pairs are duplicated? They're fine here:

```
[{"id":130,"src":"http://jywarren.github.io/Leaflet.DistortableImage/examples/example.jpg","image_file_name":"example.jpg","nodes":[
{"lat":51.5,"lng":-0.13},
{"lat":51.5,"lng":-0.09},
{"lat":51.48,"lng":-0.14},
{"lat":51.48,"lng":-0.1}
],"cm_per_pixel":44.72168905950096}]
```

But oddly they're not right here:

```
convert -contrast-stretch 0 public/warps/1562018964-working/w130-example.jpg -crop 3575x3575+0+0! -flatten -distort Perspective '0,0 0,0 0,0 0,0 0,0 0,3575 0,0 0,3575' -flatten -crop 3575x3575+0+0! +repage public/warps/1562018964/w130.png
```

Spaced out, these look like:

```
0,0 0,0
0,0 0,0
0,0 0,3575
0,0 0,3575
```

That's odd. Let me look at our logic.

**Comment by jywarren** [comment]
I reproduced this set of coordinates exactly, locally. Looking into it...

**Comment by jywarren** [comment]
OK, i was able to get this command to run locally by adding `height` and `width` to the JSON for each image:

```
convert -contrast-stretch 0 public/warps/test-working/w130-example.jpg -crop 21936x21936+0+0! -flatten -distort Perspective '0,0 0,0  521,0 0,0  521,34
8 0,17867  0,348 0,21936' -flatten -crop 21936x21936+0+0! +repage public/warps/test/w130.png
```

This seems right!

**Comment by jywarren** [comment]
And likewise in Google Cloud log: 

```
convert -contrast-stretch 0 public/warps/1562021915-working/w130-example.jpg -crop 4387x4387+0+0! -flatten -distort Perspective '0,0 0,0 521,0 0,0 521,348 0,3573 0,348 0,4387' -flatten -crop 4387x4387+0+0! +repage public/warps/1562021915/w130.png
```

**Comment by sashadev-sky** [comment]
@jywarren ok pulling in these changes now on my gh-pags to test! 

In case you needed this these are the two lines that fix the Travis error above - it comes up every once in a while
```
$ rm -rf package-lock.json node_modules
$ npm install
```

**Comment by sashadev-sky** [comment]
@jywarren its working now from this branch! But it's not completing the export. Stuck on warping 1/1. Was it completing for you / is that because i'm not on gh-pages? 

Pushing your work back up to fix the travis error and a { error that jshint is calling out

**Comment by jywarren** [comment]
I found a key issue with the exporter! It was assuming all images were either in amazon AWS or local. I've generalized it and it should now work for any image on the internet, but it will take some minutes to republish into the cloud cluster...

**Comment by jywarren** [comment]
https://github.com/publiclab/mapknitter-exporter/commit/c75c31d43b0d834c8198d43f857f4483ed5ce534 and https://github.com/publiclab/mapknitter-exporter/commit/71af895605ebba402c906af31082d6b45d81821a

**Comment by jywarren** [comment]
So, when it starts working, we'll see: `"gem_version":"1.0.9"` instead of `"gem_version":"1.0.7"`?

**Comment by jywarren** [comment]
OK, yeah it's building now...

**Comment by jywarren** [comment]
OK! Getting `1.0.9` now!

**Comment by jywarren** [comment]
OK, so let's just merge this as soon as we get it completing an export... i'm working on it!

then after that, our checklist would be:

* [ ] show a spinner, and hide it when complete!
* [ ] use `prompt('Choose a scale in real-world centimeters per pixel, or click OK to use our best guess. Keep this window open to wait for your export to complete.', this._getAvgCmPerPixels())` (i think?) to prompt for a scale, but default to the average size?
* [ ] open an issue for the "use the updated `status_url` value" which i tried to implement but it's not working

Anything else?

**Comment by jywarren** [comment]
I think we have a coordinate ordering issue:

```
 textPayload:  "convert -contrast-stretch 0 public/warps/1562024602-working/w163-example.jpg -crop 4404x4404+0+0! -flatten -distort Perspective '0,0 0,715  521,0 0,0  521,348 0,3573  0,348 0,4404' -flatten -crop 4404x4404+0+0! +repage public/warps/1562024602/w163.png

convert-im6.q16: invalid argument for option Perspective : 'Unsolvable Matrix' @ error/distort.c/GenerateCoefficients/837.
```

**Comment by jywarren** [comment]
Comparing ours:

```
[{"id":163,"src":"http://jywarren.github.io/Leaflet.DistortableImage/examples/example.jpg","width":521,"height":348,"image_file_name":"example.jpg","nodes":[
{"lat":51.52571910615877,"lng":-0.1269936262542615},
{"lat":51.529719857562284,"lng":-0.08318728898302653},
{"lat":51.509745369665325,"lng":-0.13314074967638592},
{"lat":51.505097641867415,"lng":-0.08627719376818278}
],"cm_per_pixel":99.13139883912052}]
```

And in https://mapknitter.org/maps/ceres--2/warpables.json :

```
"lat":"-37.7664063648","lon":"144.9828654528"
"lat":"-37.7650239004","lon":"144.9831980467"
"lat":"-37.7652020107","lon":"144.9844533205"
"lat":"-37.7665844718","lon":"144.9841207266"
```

**Comment by sashadev-sky** [comment]
@jywarren that sounds good. There is a leaflet spinner plugin library that would probably be the fastest solution to a spinner: https://github.com/makinacorpus/Leaflet.Spin.

I can use that and circle back and get rid of it as I build out the spinner myself? I am only now learning about SVG and implementing via web-components is not recommended with no module support so its something I need a little time for.

After merge did you want to connect to MK then come back for the UI?

**Comment by sashadev-sky** [comment]
So would the solution be to rename `lng` to `lon`?

**Comment by jywarren** [comment]
I think we need to re-order the corners we get from this section:

https://github.com/publiclab/Leaflet.DistortableImage/blob/83e3371582e2c36ea962376deec7261466a679b8/src/DistortableImageOverlay.js#L100-L109

**Comment by jywarren** [comment]
Yes, i'm going to bump version number so it'll get picked up by MapKnitter, then we can come back to UI. Thanks!

**Comment by jywarren** [comment]
OK, i'm adding a note to the README that this library does `NW, NE, SW, SE` corners!

**Comment by jywarren** [comment]
Done and pushing to my gh-pages to test: http://jywarren.github.io/Leaflet.DistortableImage/examples/select

**Comment by jywarren** [comment]
Darn! I'm still getting `convert-im6.q16: invalid argument for option Perspective : 'Unsolvable Matrix' @ error/distort.c/GenerateCoefficients/837.` -- why? let me double-check the order...

**Comment by jywarren** [comment]
```
[{"lat":51.51300227287522,"lng":-0.12977341306395831},{"lat":51.53297533262061,"lng":-0.11981705320067705},{"lat":51.53297533262061,"lng":-0.07981995237059891},{"lat":51.50869919147977,"lng":-0.07673346321098508}
```

That's: `SW NW NE SE` -- that should be correct... in the submitted order... let me check where they're coming out on the cloud logs...

**Comment by jywarren** [comment]
```
0,0 0,3573     NW to SW
521,0 0,0      NE to NW
521,348 0,0      SE to NW
0,348 0,4343      SW to SW
```

...roughly. Main problem is that points 2 and 3 of 4 are both being sent to `0,0`!

**Comment by jywarren** [comment]
Oh gosh, you're right! The exporter service uses `lon` -- ok, changing!

**Comment by sashadev-sky** [comment]
@jywarren Ok great! I had on my todo list to add to the README a blurb about coordinate order because at one point we had them backwards!

**Comment by jywarren** [comment]
Hooray! There we go! 

https://mapknitter-exports-warps.storage.googleapis.com/1562027947/1562027947.jpg

`{"status_url":"https://mapknitter-exports-warps.storage.googleapis.com/1562027947/status.json","status":"complete","tms":"public/tms/1562027947/","geotiff":"public/warps/1562027947/1562027947.tif","zip":"public/tms/1562027947.zip","jpg":"public/warps/1562027947/1562027947.jpg","export_id":1562027947,"user_id":null,"size":"78.9256MB","width":"7388","height":"2670","start_time":"2019-07-02 00:39:06 +0000","run_time":null,"gem_version":"1.0.9","cm_per_pixel":100.0}`

There are some timing issues... or something... for example I'm still getting responses like this as well:

`{"status_url":null,"status":"starting","tms":false,"geotiff":false,"zip":false,"jpg":false,"export_id":1562027947,"user_id":null,"size":null,"width":null,"height":null,"start_time":"2019-07-02 00:39:06 +0000","run_time":null,"gem_version":"1.0.9","cm_per_pixel":null}`

**Comment by sashadev-sky** [comment]
@jywarren It is working! But I am stuck on warping 1/1 still. Is this because I'm not on gh-pages?

**Comment by jywarren** [comment]
It's a big export... just give it a few more minutes?

**Comment by jywarren** [comment]
Mine has some caching issues. If i call:

`$.ajax('http://export.mapknitter.org/id/1562027947/status.json').done(function(r){console.log(r)})`

I get `status: 'starting'` even though I've /also/ received a `complete` response earlier. I keep getting `started` and so it won't kill our interval timer...

Can't seem to bust that cached response with this either:

```
$.ajax('http://export.mapknitter.org/id/1562027947/status.json?_=12421', {cache: false}).done(function(r){console.log(r)})
```

OK. We basically got there. Let's see if we can fix this cache issue and tomorrow we can merge this and publish it.

**Comment by jywarren** [comment]
Ah! OK! So -- here it is -- if we try to bust the cache by appending a timestamp, like:

`$.ajax('http://export.mapknitter.org/id/1562027947/status.json?123').done(function(r){console.log(r)})`

It will strip the `?123` when it redirects to `https://mapknitter-exports-warps.storage.googleapis.com/1562027947/status.json`

But if we directly call the final URL with a `?123` appended to bust /that/ cache, it works! 

`$.ajax('https://mapknitter-exports-warps.storage.googleapis.com/1562027947/status.json?123').done(function(r){console.log(r)})`

So, we really do have to get this segment of code working: 

https://github.com/publiclab/Leaflet.DistortableImage/blob/2d7ad24ecf72acd5ea3ee4d80a1b13721dce7d97/src/DistortableCollection.js#L353-L359

Where it can detect the "final" address.

**Comment by jywarren** [comment]
Actually I think I can fix it here too:

https://github.com/publiclab/mapknitter-exporter-sinatra/blob/4620dd9d28edae612fe59ee30a1ed303f8e47289/app.rb#L51-L59

OK - i did this in that project. I'm going to merge now! Let's do all this follow-up too!

**Comment by jywarren** [comment]
So, prompt, spinner, and then MK integration!

**Comment by jywarren** [comment]
Exciting!

**Comment by jywarren** [comment]
and -- we should rejigger the section of code above that tries to match "complete" and the new JSON address. I think if we fix it, things will run a lot smoother!

**Comment by jywarren** [comment]
I published this to NPM!

**Comment by sashadev-sky** [comment]
@jywarren should I try to connect it in MK first?

**Comment by sashadev-sky** [comment]
Ill open a new issue for the next 3 parts u just listed!

**Comment by jywarren** [comment]
I'm calling it a night! I'll check in in the morning, thanks Sasha!

**Pull request merged**